src/chipper/keyboard2.cljs

```diff
@@ -0,0 +1,233 @@
+(ns chipper.keyboard2 ; electric booga--[being dragged away by the deep state]
+  (:require [chipper.chips :as c]
+            [chipper.utils :as u]
+            [goog.events :as events]))
+
+;; ok so here's the plan
+;; event -> chan (queue) -> process (produce something) -> chan (queue) -> handler
+;; "something" is probably just a vec? or a map?
+;; like [property parameters]
+;; property and parameters would be like
+;; :position  [dline dchan dattr]  ; let's say numerical values + :first :last
+;; :attribute [line chan attr val] ; attr is any of the three note position
+;; :mode      :edit                ; :normal :edit :visual
+;; (and it's called edit because it behaves like replace, which doesn't make sense
+;; as a name in the absence of a distinct insert mode)
+;; :special ; which is then its own thing, like {:action :octave-up :parameters 10}
+;; :custom  ; or maybe `:macro`? something to facilitate more complex things
+;; then a big handler decides which little handler actually alters state
+;; big handler would basically just be a another map like
+;; {:position  set-position!
+;;  :attribute set-attribute!  ; this needs a better name than attribute, maybe
+;;  etc}
+
+;; Available properties are
+;; :position :frame :mode :attr :macro :special
+;; Available macro types are
+;; :simple   (a sequence of position/frame/mode/attr)
+;; :advanced (a callable?)
+;; and macros mappings are defined as
+;; {:Trigger [:macro {:type type :value value}]}
+;; See edit mode :Backspace for an example
+
+(def measure-size 4)
+
+;; XXX These should maybe be somwhere else, like core, or... something
+;; also note that if you use these, the max index of the cursor is 1 less
+;; because position is 0-indexed
+(def frame-count 32)
+(def frame-length 32)
+(def chan-count 4)
+(def attr-count 3)
+;; a hack, because of course
+(def total-attr-positions (* chan-count attr-count))
+
+(def -motions
+  {:down  [ 1  0  0]
+   :up    [-1  0  0]
+   :left  [ 0  0 -1]
+   :right [ 0  0  1]})
+
+(def normal-keymap
+  {;; relative movement
+   ;; keeping :position for clarity. The representation for cursor position may
+   ;; change in the future, thus -motions
+   :KeyJ        [:relative-position (:down  -motions)]
+   :KeyK        [:relative-position (:up    -motions)]
+   :KeyH        [:relative-position (:left  -motions)]
+   :KeyL        [:relative-position (:right -motions)]
+
+   :KeyEnter    [:relative-position (:down  -motions)]
+   :ArrowDown   [:relative-position (:down  -motions)]
+   :ArrowUp     [:relative-position (:up    -motions)]
+   :ArrowLeft   [:relative-position (:left  -motions)]
+   :ArrowRight  [:relative-position (:right -motions)]
+
+   :KeyW        [:relative-position [measure-size     0 0]]
+   :KeyB        [:relative-position [(- measure-size) 0 0]]
+
+   :ShiftBracketRight [:frame  1]
+   :ShiftBracketLeft  [:frame -1]
+
+   ;; absolute movement
+   :KeyG        [:absolute-position [:first :first :first]]
```

Comment:
These could actually be relative too. Just set the change to the extreme values, i.e. the constants at the top of this file (which should be put somewhere else!)

So `$` could be `[frame-length 0 0 ]`.

Then there's no need for `set-absolute-position!` to do any translation.

Reply:
I wonder if there's a way to get `ReaderT` like semantics in clojure, where everything in this file uses the symbol `:first` etc. but that gets defined somewhere in the app config page. 

when defining the keymapping, `:first`, `:last` etc. are all abstract, but we can still make all position stuff absolute by making keymaps a function from "relative position values" to "a complete keymap for this app". Then keymaps are just all defined in terms of these "relative position values" and very rarely do you see naked values.

I mean obviously we could just use a `ReaderT` implementation but we could probably get away without that overhead.

Reply:
Sure, I was looking for a way to move the constants in this file out somehow anyway. I'll shovel them and the rest of these bare values into some config map at the top of this file for now, leave _real_ config out of scope for this particular change.